src/OrleansRuntime/MembershipService/MembershipTableFactory.cs

```diff
@@ -0,0 +1,115 @@
+using System;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Orleans.Runtime.Configuration;
+
+namespace Orleans.Runtime.MembershipService
+{
+    internal class MembershipTableFactory
+    {
+        private readonly IServiceProvider serviceProvider;
+        private readonly AsyncLock initializationLock = new AsyncLock();
+        private readonly Logger logger;
+        private IMembershipTable membershipTable;
+
+        public MembershipTableFactory(IServiceProvider serviceProvider)
+        {
+            this.serviceProvider = serviceProvider;
+            logger = LogManager.GetLogger(nameof(MembershipTableFactory), LoggerType.Runtime);
+        }
+
+        internal async Task<IMembershipTable> GetMembershipTable()
+        {
+            if (membershipTable != null) return membershipTable;
+            using (this.initializationLock.LockAsync())
+            {
+                if (membershipTable != null) return membershipTable;
```

Comment:
It's also called from `ManagementGrain`, but I agree that the lock should have zero contention. It was done for correctness even though the current usage pattern doesn't require it. I can change it if you wish